src/opf/dcp.jl

```diff
@@ -1,263 +1,179 @@
-        model[:voltage_difference_limit][i] = JuMP.@constraint(model, branch["angmin"] <= va_fr - va_to <= branch["angmax"])
+        model[:va_diff][e] = @constraint(model,
+            dvamin[e] ≤ branch_status[e] * (va[bus_fr[e]] - va[bus_to[e]]) ≤ dvamax[e]
+        )
     end
 
     #
     #   III. Objective
     #
-    l, u = extrema(gen["cost"][1] for (i, gen) in ref[:gen])
-    (l == u == 0.0) || @warn "Data $(data["name"]) has quadratic cost terms; those terms are being ignored"
-    JuMP.@objective(model, Min, sum(
-        gen["cost"][2]*pg[i] + gen["cost"][3]
-        for (i,gen) in ref[:gen]
-    ))
-
-    return OPFModel{DCOPF}(data, model)
-end
+    l, u = extrema(c2[g] for g in 1:G if gen_status[g])
+    (l == u == 0.0) || @warn "Data $(data.case) has quadratic cost terms; those terms are being ignored"
 
+    @objective(model,
+        Min,
+        sum(
+            gen_status[g] * (c1[g] * pg[g] + c0[g]) for g in 1:G
+        )
+    )
 
-"""
-    _extract_dcopf_solution(model, data)
+    # TODO: update to store OPFData when all formulations are done
+    return OPFModel{DCOPF}(network, model)
+end
 
-Extract DCOPF solution from optimization model.
-The model must have been solved before.
-"""
-function extract_result(opf::OPFModel{DCOPF})
-    data  = opf.data
+function extract_primal(opf::OPFModel{DCOPF})
     model = opf.model
 
-    # Pre-process data
-    ref = PM.build_ref(data)[:it][:pm][:nw][0]
-    N = length(ref[:bus])
-    G = length(ref[:gen])
-    E = length(data["branch"])
-
-    # Build the solution dictionary
-    res = Dict{String,Any}()
-    res["opf_model"] = string(model.ext[:opf_model])
-    res["objective"] = JuMP.objective_value(model)
-    res["objective_lb"] = try JuMP.dual_objective_value(model) catch; NaN end
-    res["optimizer"] = JuMP.solver_name(model)
-    res["solve_time"] = JuMP.solve_time(model)
-    res["termination_status"] = JuMP.termination_status(model)
-    res["primal_status"] = JuMP.primal_status(model)
-    res["dual_status"] = JuMP.dual_status(model)
-    res["solution"] = sol = Dict{String,Any}()
-
-    sol["per_unit"] = get(data, "per_unit", false)
-    sol["baseMVA"]  = get(data, "baseMVA", 100.0)
-
-    ### populate branches, gens, buses ###
-
-    sol["bus"] = Dict{String,Any}()
-    sol["branch"] = Dict{String,Any}()
-    sol["gen"] = Dict{String,Any}()
-
-    for bus in 1:N
-        sol["bus"]["$bus"] = Dict(
-            "va" => value(model[:va][bus]),
-            # dual vars
-            "lam_kirchhoff" => dual(model[:kirchhoff][bus])
-        )
-    end
+    # TODO: remove when all formulations are done
+    network = opf.data
+    data = OPFData(network)
+
+    N, E, G = data.N, data.E, data.G
+
+    primal_solution = Dict{String,Any}(
+        "va" => zeros(Float64, N),
+        "pg" => zeros(Float64, G),
+        "pf" => zeros(Float64, E),
+    )
+    if has_values(model)
+        for i in 1:N
+            primal_solution["va"][i] = value(model[:va][i])
+        end
```

Comment:
I just realized that we can vectorize (broadcast) this:
```suggestion
        primal_solution["va"] .= value.(model[:va])
```
also something that affects other formulations, so we can address it all at once in #118